### ISSUE TITLE
Bug: Abbo aanzetten zorgt voor overload

### ISSUE THREAD

**Issue by thervh70** [csrdelft/csrdelft.nl#85]
Het maaltijdabonement aanzetten kan ervoor zorgen dat je als 101e persoon in een maaltijd terecht komt van een maaltijd met max100.

**Comment by Boudewijn26** [comment]
Is dit voor hele lichtingen of ook voor enkele leden?

**Comment by thervh70** [comment]
Ik zette het abonnement van een nano aan en toen zag ik het gebeuren. Weet niet hoe het zit met lichtingen.

**Comment by Boudewijn26** [comment]
Maar leden kunnen het toch niet kapot maken?

**Comment by qurben** [comment]
Kwam ik ook tegen. Moet even gecheckt worden of normale leden dit ook kunnen, anders is het een feature.

Citaat van am. Liu: Gewoon overboeken, doen professionele vliegtuigmaatschappijen ook. Prima kassen.

**Comment by qurben** [comment]
De vraag is, is het mogelijk om als een maaltijd vol zit je alsnog aan te melden door een abo aan te zetten?

**Comment by Boudewijn26** [comment]
Vraag me nu eigenlijk wel af waarom we überhaupt abbo's van sjaars aanzetten.

**Comment by qurben** [comment]
azg

**Comment by Boudewijn26** [comment]
Dat gold ook ooit voor slavernij

**Comment by qurben** [comment]
Ik vind het wel goed, motiveert ook een beetje om te komen, terwijl er de vrijheid is om af te melden.

Nu is het de vraag 'Ga ik niet?' ipv 'Ga ik wel?'

**Comment by Boudewijn26** [comment]
Ik vind het vrij paternalistisch. Vrije keuze enzo. Sjaars moeten de keuze krijgen om prominent te worden, anders is het niet hun eigen verdienste.

**Comment by qurben** [comment]
Ja, laten we ook het novitiaat afschaffen. Als sjaarsch echt prominent willen worden komen ze wel naar de borrel.

**Comment by Boudewijn26** [comment]
Ik ben voor!

**Comment by thervh70** [comment]
laten we het serieus houden hier heren.

**Comment by Boudewijn26** [comment]
Ik heb het getest en dit geld ook als leden (of ik ieder geval ik) hun abbo aanzetten.

**Comment by brussee** [comment]
> Ik heb het getest en dit geldt ook als leden (of in ieder geval ik) hun abbo aanzetten.

Dat is dan een bug, want voor beheerders is het een feature (ja serieus, heb ik aan gedacht bij ontwerp en implementatie...)

**Comment by brussee** [comment]
Het gebeurd niet als de maaltijd gesloten is (ook niet bij beheerders).

**Comment by brussee** [comment]
Het probleem zit hier: deze functie wordt vanuit de beheer-omgeving én door normale leden aangeroepen. De check of de maaltijd vol is kan je dus niet in deze functie zetten. Dit moet dus ge-refactored worden. De core van het MaalCie systeem is geschreven voordat het framework er was. Als je dat gaat gebruiken is het fixen van deze bug eenvoudig, want dit soort complexe SQL queries worden geabstraheerd.
```
private static function newAanmelding($mid, $uid, $gasten, $opmerking, $doorAbo, $doorUid) {
		$sql = 'INSERT IGNORE INTO mlt_aanmeldingen';
		$sql.= ' (maaltijd_id, uid, aantal_gasten, gasten_eetwens, door_abonnement, door_uid, laatst_gewijzigd)';
		$wanneer = date('Y-m-d H:i');
		if ($mid === null) { // niet voor specifieke maaltijd? dan voor alle komende repetitie-maaltijden
			$sql.= ' SELECT maaltijd_id, ?, ?, ?, ?, ?, ? FROM mlt_maaltijden';
			$sql.= ' WHERE mlt_repetitie_id = ? AND gesloten = FALSE AND verwijderd = FALSE AND datum >= ?';
			$values = array($uid, $gasten, $opmerking, $doorAbo, $doorUid, $wanneer, $doorAbo, date('Y-m-d'));
		} else {
...
```

**Comment by brussee** [comment]
Als je beheerders niet meer via een abonnement mensen laat aanmelden bij volle maaltijden (kan nog steeds individueel) is de volgende fix een idee: maar de query moet nog veel complexer worden dan dit, want je moet uitrekenen hoeveel aanmeldingen er nog zullen gaan plaatsvinden met de INSERT IGNORE om uit te rekenen of er nog ruimte is in de betreffende maaltijd. Oftewel: refactor naar het framework, want dan is het triviaal, zie groepen aanmeld-systeem.
```
            $values = array($uid, $gasten, $opmerking, $doorAbo, $doorUid, $wanneer, $doorAbo, date('Y-m-d'));
            if ($doorAbo) {
                    $sql.= ' AND aanmeld_limiet > SELECT COUNT(1) FROM mlt_aanmeldingen';
            }
        } else {
```
